Initial AWS Lambda documentation (#13245)
hashicorp/consul: website/content/docs/lambda/invocation.mdx

The following diagram shows the invocation procedure:

<ImageConfig width={400}>

![Connect Proxy to Lambda](/img/connect_proxy_to_lambda.svg)

</ImageConfig>

1. Make an HTTP request to the local Connect proxy.
2. The Connect proxy invokes the Lambda.

## Invoke a Lambda Function
Before you can invoke a Lambda function, register the service used to invoke
the Lambda function and the service running in Lambda with
Consul (refer to [registration](/docs/lambda/registration) for
instructions). The service used to invoke the function must be deployed to the
service mesh.

1. Update the invoking service to use the Lambda service as an upstream. In the following example, the `destination_name` for the invoking service (`api`) points to a Lambda service called `authentication`:
    ```hcl
    upstreams {
      local_bind_port = 2345
      destination_name = "authentication"
    }
    ```
1. Issue the `consul services register` command to store the configuration:
    ```shell-sesion
     $ consul services register api-sidecar-proxy.hcl`.
    ```

> 1. Call `curl localhost:2345` from `api` to invoke the `authentication` service.

Reviewer: ```suggestion
1. Call the upstream service to invoke the Lambda function. In the following example, the `api` service invokes the `authentication` service at `localhost:2345`:
    ```shell-session
     $ curl https://localhost:2345
    ```
```
Is this correct? My only concern is that we are describing how to invoke the specific example instead of providing general instructions.

are unsupported. Therefore, we recomend invoking Lambdas through terminating
gateways.

## Invocation Types

### Recommended: Terminating Gateway

Invoking Lambdas through terminating gateways supports cross-datacenter communication, transparent
proxies, intentions, and all other Consul service mesh features. It requires
the terminating gateway to have [the appropriate IAM permissions](/docs/lambda/requirements#allow-envoy-to-invoke-the-lambda)
to invoke the Lambda.

**Architecture**:

<ImageConfig width={700}>

![Terminating Gateway to Lambda](/img/terminating_gateway_to_lambda.svg)

</ImageConfig>

1. Make an HTTP request to the local Connect proxy.
2. The Connect proxy forwards the request on to the terminating gateway.
3. The terminating gateway invokes the Lambda.

### Connect Proxy

This invocation type directly invokes Lambdas from a service's Connect proxy.
It has the following limitations:
- Intentions are unsupported - Consul enforces intentions on receiving end of
  connections by validating client certificates. Lambda doesn't support
  validating client certificates, so we are unable to support intentions.

> - Transparent proxies are unsupported - Since Lambda services are not
>   registered to a proxy, so transparent proxying to not be supported.

Reviewer 2: fix phrasing:

```suggestion
- Transparent proxies are unsupported - Lambda services are not
  registered to a proxy, so transparent proxying is not supported.
```

This approach is secure because Lambdas cannot be invoked without the correct AWS
IAM permissions and invocations use Amazon’s TLS. The Envoy sidecar proxy
will need to have the correct AWS IAM credentials defined to invoke the Lambda.
These credentials can be defined in environment variables, EC2 metadata or ECS
task metadata.

**Architecture**:

<ImageConfig width={400}>

![Connect Proxy to Lambda](/img/connect_proxy_to_lambda.svg)

</ImageConfig>

1. Make an HTTP request to the local Connect proxy.
2. The Connect proxy invokes the Lambda.

## Invoking the Lambda

### Assumptions

> * A service named `api` is registered and running on Consul service mesh.

Reviewer 2: Something introductory here,

```suggestion
This section gives an example of invoking a Lambda function. In this example, we have the following services:

* A service named `api` is registered and running on Consul service mesh.
```

* A Lambda service named `authentication` [is registered into Consul](/docs/lambda/registration/overview).

### Use the Lambda Service as an Upstream

Update the `api` service's proxy registration configuration to use the
`authentication` service as an upstream:

```hcl
upstreams {
  local_bind_port = 2345
  destination_name = "authentication"
}
```

Store the configuration with `consul services register api-sidecar-proxy.hcl`.

### Invoke the Lambda Service

> Call `curl localhost:2345` from `api` to invoke the `authentication` service.

Reviewer 2: Connecting this to the previous section a little, and concluding

```suggestion
Now that the Lambda is configured with the `authentication` service, you can run `curl localhost:2345` from `api` to invoke the `authentication` service. This invokes the Lambda function associated with the `authentication` service.
```
